### PR TITLE
Additional FreeBSD fixes

### PR DESCRIPTION
**Pull request by michael-yuji**
Add FreeBSD support for swift-testing.

### Result:
With this PR, swift-testing can be built on FreeBSD (tested on x86_64 FreeBSD 14.1-RELEASE-p6)

### Known issue:
There are some issue running `swift test` on this repo, but most likely due to bugs in my host toolchain.
- Tests failed to link due to duplicated main symbols:
```
error: link command failed with exit code 1 (use -v to see invocation)
ld: error: duplicate symbol: main
>>> defined at TestingMacrosMain.swift:0 (/zdata/swift-main/swift-project/swift-testing/Sources/TestingMacros/TestingMacrosMain.swift:0)
>>>            /zdata/swift-main/swift-project/swift-testing/.build/x86_64-unknown-freebsd14.1/debug/TestingMacros-tool.build/TestingMacrosMain.swift.o:(main)
>>> defined at runner.swift:0 (/zdata/swift-main/swift-project/swift-testing/.build/x86_64-unknown-freebsd14.1/debug/swift-testingPackageTests.derived/runner.swift:0)
>>>            /zdata/swift-main/swift-project/swift-testing/.build/x86_64-unknown-freebsd14.1/debug/swift_testingPackageTests-tool.build/runner.swift.o:(.text.main+0x0)
clang: error: linker command failed with exit code 1 (use -v to see invocation)
[25/26] Linking swift-testingPackageTests.xctest
```

- [x] Code and documentation should follow the style of the [Style Guide](https://github.com/apple/swift-testing/blob/main/Documentation/StyleGuide.md).
- [x] If public symbols are renamed or modified, DocC references should be updated.

### CODE REVIEW

File: Sources/Testing/ExitTests/SpawnProcess.swift [hunk 1]
```diff
@@ -143,12 +143,14 @@ func spawnExecutable(
...
       flags |= CShort(POSIX_SPAWN_CLOEXEC_DEFAULT)
-#elseif os(Linux) || os(FreeBSD)
+#elseif os(Linux)
       // This platform doesn't have POSIX_SPAWN_CLOEXEC_DEFAULT, but we can at
       // least close all file descriptors higher than the highest inherited one.
       // We are assuming here that the caller didn't set FD_CLOEXEC on any of
       // these file descriptors.
       _ = swt_posix_spawn_file_actions_addclosefrom_np(fileActions, highestFD + 1)
+#elseif os(FreeBSD)
+      _ = posix_spawn_file_actions_addclosefrom_np(fileActions, highestFD + 1)
```
Comment (grynspan) [inline Sources/Testing/ExitTests/SpawnProcess.swift:153]: Can you explain why this change was needed?

File: Sources/Testing/ExitTests/WaitFor.swift [hunk 2]
```diff
@@ -85,7 +85,11 @@ private let _childProcessContinuations = Locked<[pid_t: CheckedContinuation<Exit
 /// A condition variable used to suspend the waiter thread created by
 /// `_createWaitThread()` when there are no child processes to await.
 private nonisolated(unsafe) let _waitThreadNoChildrenCondition = {
+  #if os(FreeBSD) || os(OpenBSD)
```
Comment (grynspan) [inline Sources/Testing/ExitTests/WaitFor.swift:88]:
```suggestion
  #if os(FreeBSD)
```

File: Sources/Testing/Support/Locked.swift [hunk 3]
```diff
@@ -149,6 +151,34 @@ struct Locked<T>: RawRepresentable, Sendable where T: Sendable {
       }
     }
   }
+#elseif os(FreeBSD)
```
Comment (grynspan) [inline Sources/Testing/Support/Locked.swift:154]: Rather than duplicating this entire function, why not just change the argument type to `UnsafeMutablePointer<_Lock>`?

File: Sources/_TestingInternals/include/Includes.h [hunk 4]
```diff
@@ -80,6 +80,10 @@
 #include <pthread.h>
 #endif
 
+#if __has_include(<pthread_np.h>)
+#include <pthread_np.h>
```
Comment (grynspan) [inline Sources/_TestingInternals/include/Includes.h:84]: I assume this header contains `posix_spawn_file_actions_addclosefrom_np()`?

File: Sources/Testing/ExitTests/ExitTest.swift [hunk 5]
```diff
@@ -114,7 +114,7 @@ extension ExitTest {
     // As with Linux, disable the generation core files. FreeBSD does not, as
     // far as I can tell, special-case RLIMIT_CORE=1.
     var rl = rlimit(rlim_cur: 0, rlim_max: 0)
-    _ = setrlimit(CInt(RLIMIT_CORE.rawValue), &rl)
+    _ = setrlimit(CInt(RLIMIT_CORE), &rl)
```
Comment (grynspan) [inline Sources/Testing/ExitTests/ExitTest.swift:117]: If you remove the cast to `CInt`, does this line compile?

**Comment by grynspan** [comment]
Thanks for taking a look at this! I've left you some questions.

**Comment by grynspan** [comment]
> Some tests failed to build (due to missing openpty)

FreeBSD has the [`openpty()`](https://man.freebsd.org/cgi/man.cgi?query=openpty&sektion=3) function. Is there a missing `#include` in _TestingInternals that we can add?

File: Sources/Testing/ExitTests/SpawnProcess.swift (hunk 1 above)
Comment (michael-yuji) [reply Sources/Testing/ExitTests/SpawnProcess.swift:153]: FreeBSD do not use Glibc, and all non-EOL (13.1+) FreeBSD libc [supports `posix_spawn_file_actions_addclosefrom_np`](https://man.freebsd.org/cgi/man.cgi?query=posix_spawn_file_actions_addclosefrom_np&apropos=0&sektion=0&manpath=FreeBSD+14.2-RELEASE+and+Ports&arch=default&format=html).

`swt_posix_spawn_file_actions_addclosefrom_np` on the other hand is defined in `Stubs.cpp` to guard the actual implementation behind `#if defined(__GLIBC__)` and `__GLIBC_PREREQ(2, 34)`, such that Linux host with lower glibc version can still works (and becomes no-op). And changing `Stubs.cpp` will end up with a less trivial diff.

File: Sources/_TestingInternals/include/Includes.h (hunk 4 above)
Comment (michael-yuji) [reply Sources/_TestingInternals/include/Includes.h:84]: yup

File: Sources/Testing/Support/Locked.swift (hunk 3 above)
Comment (michael-yuji) [reply Sources/Testing/Support/Locked.swift:154]: `_Lock` is defined as private and therefore cannot be use here as this function is internal.

File: Sources/Testing/ExitTests/SpawnProcess.swift (hunk 1 above)
Comment (grynspan) [reply Sources/Testing/ExitTests/SpawnProcess.swift:153]: Of course. :) Mind adding a brief comment above the function call so folks reading the code know why it's different?

File: Sources/Testing/ExitTests/ExitTest.swift (hunk 5 above)
Comment (michael-yuji) [reply Sources/Testing/ExitTests/ExitTest.swift]: it does, changed accordingly.

File: Sources/Testing/Support/Locked.swift (hunk 3 above)
Comment (grynspan) [reply Sources/Testing/Support/Locked.swift:154]: The typealias can be made `internal` (no keyword) if needed.

File: Sources/Testing/ExitTests/SpawnProcess.swift (hunk 1 above)
Comment (michael-yuji) [reply Sources/Testing/ExitTests/SpawnProcess.swift:158]: ofc! done!

**Comment by grynspan** [comment]
@swift-ci test

**Comment by michael-yuji** [comment]
> > Some tests failed to build (due to missing openpty)
> 
> FreeBSD has the [`openpty()`](https://man.freebsd.org/cgi/man.cgi?query=openpty&sektion=3) function. Is there a missing `#include` in _TestingInternals that we can add?

This is the weird part. `openpty` supposed to be in `libutil.h`, but for some reason even if I added `libutil.h` to _TestingItnernals Includes.h the compiler still couldn't see the symbol and I don't know why.

**Comment by grynspan** [comment]
@swift-ci test

**Review by grynspan — APPROVED** [review]

**Comment by grynspan** [comment]
@swift-ci test

**Comment by grynspan** [comment]
Related: https://github.com/swiftlang/swift/pull/77836

**Comment by michael-yuji** [comment]
Related: https://github.com/swiftlang/swift-package-manager/pull/8193
The above swiftpm PR should've fix the duplicated symbol error for Functional Tests.
